peachjam/views/documents.py

```diff
@@ -9,6 +11,55 @@
 from peachjam.registry import registry
 
 
+class RedirectResolver:
+
+    RESOLVER_MAPPINGS = {
+        "africanlii": {"code": "aa", "domain": "agp.africanlii.org"},
+        "ghalii": {"code": "gh", "domain": "new.ghalii.org"},
+        "lawlibrary": {"code": "za", "domain": "lawlibrary.org.za"},
+        "lesotholii": {"code": "ls", "domain": "new.lesotholii.org"},
+        "malawilii": {"code": "mw", "domain": "malawilii.org"},
+        "namiblii": {"code": "na", "domain": "namiblii.org"},
+        "senlii": {"code": "sn", "domain": "senlii.org"},
+        "seylii": {"code": "sc", "domain": "new.seylii.org"},
+        "sierralii": {"code": "sl", "domain": "new.sierralii.org"},
+        "tanzlii": {"code": "tz", "domain": "new.tanzlii.org"},
+        "tcilii": {"code": "tc", "domain": "new.tcilii.org"},
+        "ulii": {"code": "ug", "domain": "new.ulii.org"},
+        "zambialii": {"code": "zm", "domain": "new.zambialii.org"},
+        "zanzibarlii": {"code": "tz-znz", "domain": "new.zanzibarlii.org"},
+        "zimlii": {"code": "zw", "domain": "zimlii.org"},
+    }
+
+    def __init__(self):
+        self.current_authority = self.RESOLVER_MAPPINGS[
+            settings.PEACHJAM["APP_NAME"].lower()
+        ]
+
+    def get_url_for_frbr_uri(self, frbr_uri):
+        try:
+            parsed_uri = FrbrUri.parse(frbr_uri)
+            country_code = parsed_uri.country
+            locality = parsed_uri.locality
+
+            # special case for tanzlii and zanzibarlii. All others are authoritative for their countries
+            if (
+                country_code == "tz"
+                and locality == "znz"
+                and self.current_authority["domain"]
+                != self.RESOLVER_MAPPINGS["zanzibarlii"]["domain"]
+            ):
+                return f"http://{self.RESOLVER_MAPPINGS['zanzibarlii']['domain']}{frbr_uri}"
```

Comment:
You're building the final URL in two places, which means you're duplicating code.

Rather, split this into two steps:

1. identify which authority to use (if any) -- handle any authority-lookup edge cases here
2. based on that authority, generate a URL

Reply:
I'd say splitting this method into two separate methods would make that easier.

Reply:
I've split the method up into two, one for getting best match and one for comparing with current authority